auto_compressor/src/state_saving.rs

```diff
@@ -0,0 +1,260 @@
+// This module contains functions to communicate with the database
+
+use anyhow::{bail, Result};
+use core::fmt;
+use rand::{distributions::Alphanumeric, thread_rng, Rng};
+use std::borrow::Cow;
+use synapse_compress_state::Level;
+
+use openssl::ssl::{SslConnector, SslMethod, SslVerifyMode};
+use postgres::{fallible_iterator::FallibleIterator, Client};
+use postgres_openssl::MakeTlsConnector;
+
+/// Connects to the database and returns a postgres client
+///
+/// # Arguments
+///
+/// * `db_url`          -   The URL of the postgres database that synapse is using.
+///                         e.g. "postgresql://user:password@domain.com/synapse"
+pub fn connect_to_database(db_url: &str) -> Result<Client> {
+    let mut builder = SslConnector::builder(SslMethod::tls())?;
+    builder.set_verify(SslVerifyMode::NONE);
+    let connector = MakeTlsConnector::new(builder.build());
+
+    let client = Client::connect(db_url, connector)?;
+    Ok(client)
+}
+
+/// Creates the state_compressor_state and state_compressor progress tables
+///
+/// If these tables already exist then this function does nothing
+///
+/// # Arguments
+///
+/// * `client`        - A postgres client used to send the requests to the database
+pub fn create_tables_if_needed(client: &mut Client) -> Result<()> {
+    let create_state_table = r#"
+        CREATE TABLE IF NOT EXISTS state_compressor_state (
+            room_id TEXT NOT NULL,
+            level_num INT NOT NULL,
+            max_size INT NOT NULL,
+            current_length INT NOT NULL,
+            current_head BIGINT,
+            UNIQUE (room_id, level_num)
+        )"#;
+
+    client.execute(create_state_table, &[])?;
+
+    let create_state_table_indexes = r#"
+        CREATE INDEX IF NOT EXISTS state_compressor_state_index ON state_compressor_state (room_id)"#;
+
+    client.execute(create_state_table_indexes, &[])?;
+
+    let create_progress_table = r#"
+        CREATE TABLE IF NOT EXISTS state_compressor_progress (
+            room_id TEXT NOT NULL,
+            last_compressed BIGINT NOT NULL,
+            UNIQUE (room_id)
+        )"#;
+
+    client.execute(create_progress_table, &[])?;
+
+    Ok(())
+}
+
+/// Retrieve the level info so we can restart the compressor
+///
+/// # Arguments
+///
+/// * `client`        - A postgres client used to send the requests to the database
+/// * `room_id`       - The room who's saved compressor state we want to load
+pub fn read_room_compressor_state(
+    client: &mut Client,
+    room_id: &str,
+) -> Result<Option<(i64, Vec<Level>)>> {
+    // Query to retrieve all levels from state_compressor_state
+    // Ordered by ascending level_number
+    let sql = r#"
+        SELECT level_num, max_size, current_length, current_head, last_compressed
+        FROM state_compressor_state as s
+        LEFT JOIN state_compressor_progress as p ON s.room_id = p.room_id
```

Comment:
Well you don't want to be able to restart the compressor if progress is NULL but there are levels start, but neither do you want to start the compressor from scratch.

Consider the following scenario:
 1: compressor is run on a room with default level sizes of 100,50,25
 2: that room's entry in state_compressor_progress is deleted by accident
 3: compressor is run on that same room with default level sizes of 100,50 (i.e. just two levels)
 
If we now start the compressor, when we save the compressor state, we don't overwrite the contents of the level 3 (the one with max size 25) so when the compressor loads state from the database it will load that third level with corrupted data in.

So I've added a `bail!` so it spots the mistake early!